Source/WebCore/animation/CSSPropertyAnimation.cpp

```diff
@@ -3910,13 +3911,42 @@ void CSSPropertyAnimation::blendProperties(const CSSPropertyBlendingClient* clie
     }
 }
 
-void CSSPropertyAnimation::blendCustomProperty(const AtomString& customProperty, RenderStyle& destination, const RenderStyle& from, const RenderStyle& to, double progress)
+template <typename T>
+static std::optional<T> typedCustomPropertyValue(const CSSCustomPropertyValue::VariantValue& variantValue)
 {
-    const auto& source = progress < 0.5 ? from : to;
-    if (auto nonInheritedValue = source.nonInheritedCustomProperties().get(customProperty))
-        destination.setNonInheritedCustomPropertyValue(customProperty, CSSCustomPropertyValue::create(*nonInheritedValue));
-    else if (auto inheritedValue = source.inheritedCustomProperties().get(customProperty))
-        destination.setInheritedCustomPropertyValue(customProperty, CSSCustomPropertyValue::create(*inheritedValue));
+    if (std::holds_alternative<CSSCustomPropertyValue::SyntaxValue>(variantValue)) {
+        auto& syntaxValue = std::get<CSSCustomPropertyValue::SyntaxValue>(variantValue);
+        if (std::holds_alternative<T>(syntaxValue))
+            return std::get<T>(syntaxValue);
+    }
+    return std::nullopt;
+}
+
+static Ref<CSSCustomPropertyValue> blendedCSSCustomPropertyValue(const CSSCustomPropertyValue& from, const CSSCustomPropertyValue& to, const CSSPropertyBlendingContext& blendingContext)
+{
+    if (auto possibleFromLength = typedCustomPropertyValue<Length>(from.value())) {
+        ASSERT(typedCustomPropertyValue<Length>(to.value()));
+        auto blendedLength = blendFunc(*possibleFromLength, *typedCustomPropertyValue<Length>(to.value()), blendingContext);
+        return CSSCustomPropertyValue::createForLengthSyntax(from.name(), WTFMove(blendedLength));
+    }
```

Comment:
This seems overly complex and hard-to-follow way to dot this. I think you should first test if `to` and `from` both contain `CSSCustomPropertyValue::SyntaxValue`. If so it would next invoke another blending function that takes `from` and `to` as `SyntaxValue`s and returns a  `SyntaxValue`. That function then does `switchOn` to the actual type and invokes the type specific being function.

Reply:
I attempted to do this in the new patch. Hopefully this is an improvement.